server/venueless/core/models/world.py

```diff
@@ -44,10 +80,68 @@ def decode_token(self, token):
                     issuer=issuer,
                 )
 
-    def has_permission(self, permission, traits, extra_config=None):
-        permission_config = copy.deepcopy(self.permission_config)
-        if extra_config:
-            permission_config.update(extra_config)
-        return permission in permission_config and all(
-            trait in traits for trait in permission_config[permission]
-        )
+    def has_permission_implicit(self, *, traits, permission: Permission, room=None):
+        for role, required_traits in self.trait_grants.items():
+            if isinstance(required_traits, list) and all(
+                r in traits for r in required_traits
+            ):
+                if permission.value in self.roles[role]:
+                    return True
+
+        if room:
+            for role, required_traits in room.trait_grants.items():
+                if isinstance(required_traits, list) and all(
+                    r in traits for r in required_traits
+                ):
+                    if permission.value in self.roles[role]:
+                        return True
+
+    def has_permission(self, *, user, permission: Permission, room=None):
+        """
+        Returns whether a user holds a given permission either on the world or on a specific room.
+        """
+        if self.has_permission_implicit(
+            traits=user.traits, permission=permission, room=room
+        ):
+            return True
+
+        roles = user.get_role_grants(room)
+        for r in roles:
+            if permission.value in self.roles[r]:
+                return True
+
+    async def has_permission_async(self, *, user, permission: Permission, room=None):
```

Comment:
Yeah :( AFAIK there's still no way to make a function callable as both one or the other. Generally, the sync_to_async/async_to_sync things help, but in this case I decided for code duplication since it's going to be used a lot – and also allows for some relevant optimizations since the function only needs the database if the implicit grants do not already give permission.